src/client/opamClient.ml

```diff
@@ -315,6 +315,63 @@ let upgrade_t ?strict_upgrade ?auto_install ?ask ?(check=false) ~all atoms t =
              (OpamPackage.names_of_packages t.compiler_packages)
          in
          let unopt = unopt -- base in
+         let get_formula pkg get =
+           OpamStd.Option.map
+             (fun opam ->
+                OpamFilter.filter_formula ~default:false
+                  (OpamFilter.deps_var_env ~build:true ~post:false
+                     ~test:true ~dev:false ~doc:false)
+                  (get opam))
+             (OpamSwitchState.opam_opt t pkg)
+         in
+         let conflicts =
+           OpamPackage.Set.fold (fun unopt_pkg map ->
+               let set =
+                 OpamSwitchState.conflicts_with t
+                   (OpamPackage.Set.singleton unopt_pkg) latest in
+               OpamPackage.Set.fold (fun installed_pkg map1 ->
+                   match get_formula installed_pkg  OpamFile.OPAM.conflicts with
+                   | None -> map1
+                   | Some conflicts_formula ->
+                     if OpamFormula.verifies conflicts_formula unopt_pkg then
+                       OpamFormula.fold_left
+                         (fun map2 (n,formula) ->
+                            if OpamPackage.name unopt_pkg = n
+                            then
+                              OpamPackage.Map.update unopt_pkg
+                                (OpamStd.List.cons (installed_pkg, formula)) [] map2
+                            else map2
+                         ) map1 conflicts_formula
+                     else map1
```

Comment:
While not strictly correct in the case of disjunctions (you only consider atoms, so if a package has `depends: "foo" { (< "3" | > "4") & !=5 }` and you have `foo=5`, you would retain both `< "3"` and `!=5`), I think it is fine as it is (it's just a little artefact in the worst — very rare — case, and more complicated than it deserves to do better). Maybe a comment could say so, though.

Reply:
Disjunctions are handled: each package formula is checked using `OpamFormula.check_version_formula`, then added to the map for printing.